File: tests/end-to-end/ec2_encryption/csharp/Stack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.EC2;
using Constructs;
using System.Collections.Generic;

namespace Ec2EncryptionStack
{
    public class Ec2EncryptionStackProps : StackProps
    {
        public string Environment { get; set; }

        public string DatabaseType { get; set; }

        public bool? UseEncryption { get; set; }

        public string EncryptedAmi { get; set; }

        public string UnencryptedAmi { get; set; }

        public string SubnetType { get; set; }

        public bool? EnableMonitoringParameter { get; set; }

    }

    public class Ec2EncryptionStack : Stack
    {
        public Ec2EncryptionStack(Construct scope, string id, Ec2EncryptionStackProps props = null) : base(scope, id, props)
        {
            // Applying default props
            props ??= new Ec2EncryptionStackProps();
            props.Environment ??= "dev";
            props.DatabaseType ??= "postgresql";
            props.UseEncryption ??= false;
            props.EncryptedAmi ??= "ami-1234567890abcdef0";
            props.UnencryptedAmi ??= "ami-0987654321fedcba0";
            props.SubnetType ??= "Private1";
            props.EnableMonitoringParameter ??= false;

            // Mappings
            var regionToAmi = new Dictionary<string, Dictionary<string,string>> 
            {
                ["us-east-1"] = new Dictionary<string, string> {["AMI"] = "ami-12345678", },
                ["us-west-2"] = new Dictionary<string, string> {["AMI"] = "ami-87654321", },
            };

            // Conditions
            bool hasDatabase = props.DatabaseType == "mysql";
            bool isProduction = props.Environment == "prod";
            bool usePrivateSecurityGroup = props.SubnetType == "Private1" || props.SubnetType == "Private2";
            bool keyPairProd = !(isProduction);
            bool useEncryption = isProduction && hasDatabase;

            // Resources
            var privateSecurityGroup = new CfnSecurityGroup(this, "PrivateSecurityGroup", new CfnSecurityGroupProps
            {
                GroupDescription = "Private security group",
                VpcId = "vpc-xxxxxxxx",
            });
            var publicSecurityGroup = new CfnSecurityGroup(this, "PublicSecurityGroup", new CfnSecurityGroupProps
            {
                GroupDescription = "Public security group",
                VpcId = "vpc-xxxxxxxx",
            });
            var myApp = new CfnInstance(this, "MyApp", new CfnInstanceProps
            {
                ImageId = regionToAmi["us-east-1"]["AMI"],
                SecurityGroups = new []
                {
                    usePrivateSecurityGroup ? privateSecurityGroup.Ref : publicSecurityGroup.Ref,
                },
            });
        }
    }
}

File: tests/end-to-end/ec2_encryption/golang/stack.go
package main

import (
	cdk "github.com/aws/aws-cdk-go/awscdk/v2"
	ec2 "github.com/aws/aws-cdk-go/awscdk/v2/awsec2"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type Ec2EncryptionStackProps struct {
	cdk.StackProps
	Environment *string
	DatabaseType *string
	UseEncryption interface{/* Boolean */}
	EncryptedAmi *string
	UnencryptedAmi *string
	SubnetType *string
	EnableMonitoringParameter interface{/* Boolean */}
}

type Ec2EncryptionStack struct {
	cdk.Stack
}

func NewEc2EncryptionStack(scope constructs.Construct, id string, props *Ec2EncryptionStackProps) *Ec2EncryptionStack {
	regionToAmi := map[*string]map[*string]*string{
		jsii.String("us-east-1"): map[*string]*string{
			jsii.String("AMI"): jsii.String("ami-12345678"),
		},
		jsii.String("us-west-2"): map[*string]*string{
			jsii.String("AMI"): jsii.String("ami-87654321"),
		},
	}

	var sprops cdk.StackProps
	if props != nil {
		sprops = props.StackProps
	}
	stack := cdk.NewStack(scope, &id, &sprops)

	hasDatabase := props.DatabaseType == jsii.String("mysql")

	isProduction := props.Environment == jsii.String("prod")

	usePrivateSecurityGroup := props.SubnetType == jsii.String("Private1") || props.SubnetType == jsii.String("Private2")

	keyPairProd := !isProduction

	useEncryption := isProduction && hasDatabase

	privateSecurityGroup := ec2.NewCfnSecurityGroup(
		stack,
		jsii.String("PrivateSecurityGroup"),
		&ec2.CfnSecurityGroupProps{
			GroupDescription: jsii.String("Private security group"),
			VpcId: jsii.String("vpc-xxxxxxxx"),
		},
	)

	publicSecurityGroup := ec2.NewCfnSecurityGroup(
		stack,
		jsii.String("PublicSecurityGroup"),
		&ec2.CfnSecurityGroupProps{
			GroupDescription: jsii.String("Public security group"),
			VpcId: jsii.String("vpc-xxxxxxxx"),
		},
	)

	ec2.NewCfnInstance(
		stack,
		jsii.String("MyApp"),
		&ec2.CfnInstanceProps{
			ImageId: regionToAmi[jsii.String("us-east-1")][jsii.String("AMI")],
			SecurityGroups: &[]*string{
				ifCondition(
					usePrivateSecurityGroup,
					privateSecurityGroup.Ref(),
					publicSecurityGroup.Ref(),
				),
			},
		},
	)

	return &Ec2EncryptionStack{
		Stack: stack,
	}
}

/// ifCondition is a helper function that replicates the ternary
/// operator that can be found in other languages. It is conceptually
/// equivalent to writing `cond ? whenTrue : whenFalse`, meaning it
/// returns `whenTrue` if `cond` is `true`, and `whenFalse` otherwise.
func ifCondition[T any](cond bool, whenTrue T, whenFalse T) T {
	if cond {
		return whenTrue
	}
	return whenFalse
}

